storage: TestPebbleMetricEventListener failed

storage.TestPebbleMetricEventListener [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8599143?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/8599143?buildTab=artifacts#/) on release-22.2 @ [94796359a4a73690b8e39484d8ecc515b64db77b](https://github.com/cockroachdb/cockroach/commits/94796359a4a73690b8e39484d8ecc515b64db77b):


```
=== RUN   TestPebbleMetricEventListener
    test_log_scope.go:161: test logs captured to: /artifacts/tmp/_tmp/c5eb8fc8b8e683c19f3c3e4238f64094/logTestPebbleMetricEventListener1760716648
    test_log_scope.go:79: use -show-logs to present logs inline
    pebble_test.go:272: -- test log scope end --
    test_log_scope.go:412: unlinkat /artifacts/tmp/_tmp/c5eb8fc8b8e683c19f3c3e4238f64094/logTestPebbleMetricEventListener1760716648: directory not empty
--- FAIL: TestPebbleMetricEventListener (0.01s)
```
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #96635 storage: TestPebbleMetricEventListener failed [C-test-failure O-robot T-storage branch-master]
</p>
</details>
/cc @cockroachdb/storage
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestPebbleMetricEventListener.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-24257

--- a/pkg/storage/pebble.go
+++ b/pkg/storage/pebble.go
@@ -693,6 +693,7 @@ type Pebble struct {
 		syncutil.Mutex
 		flushCompletedCallback func()
 	}
+	asyncDone sync.WaitGroup
 
 	// supportsRangeKeys is 1 if the database supports range keys. It must
 	// be accessed atomically.
@@ -943,7 +944,7 @@ func NewPebble(ctx context.Context, cfg PebbleConfig) (p *Pebble, err error) {
 	oldDiskSlow := lel.DiskSlow
 	lel.DiskSlow = func(info pebble.DiskSlowInfo) {
 		// Run oldDiskSlow asynchronously.
-		go oldDiskSlow(info)
+		p.async(func() { oldDiskSlow(info) })
 	}
 	cfg.Opts.EventListener = pebble.TeeEventListener(
 		p.makeMetricEtcEventListener(ctx),
@@ -990,6 +991,17 @@ func NewPebble(ctx context.Context, cfg PebbleConfig) (p *Pebble, err error) {
 	}
 
 	return p, nil
+}
+
+// async launches the provided function in a new goroutine. It uses a wait group
+// to synchronize with (*Pebble).Close to ensure all launched goroutines have
+// exited before Close returns.
+func (p *Pebble) async(fn func()) {
+	p.asyncDone.Add(1)
+	go func() {
+		defer p.asyncDone.Done()
+		fn()
+	}()
 }
 
 func (p *Pebble) makeMetricEtcEventListener(ctx context.Context) pebble.EventListener {
@@ -1042,8 +1054,10 @@ func (p *Pebble) makeMetricEtcEventListener(ctx context.Context) pebble.EventLis
 					log.Fatalf(ctx, "disk stall detected: pebble unable to write to %s in %.2f seconds",
 						info.Path, redact.Safe(info.Duration.Seconds()))
 				} else {
-					go log.Errorf(ctx, "disk stall detected: pebble unable to write to %s in %.2f seconds",
-						info.Path, redact.Safe(info.Duration.Seconds()))
+					p.async(func() {
+						log.Errorf(ctx, "disk stall detected: pebble unable to write to %s in %.2f seconds",
+							info.Path, redact.Safe(info.Duration.Seconds()))
+					})
 				}
 				return
 			}
@@ -1082,6 +1096,9 @@ func (p *Pebble) Close() {
 		return
 	}
 	p.closed = true
+
+	// Wait for any asynchronous goroutines to exit.
+	p.asyncDone.Wait()
 
 	handleErr := func(err error) {
 		if err == nil {

